File: src/ImageViewer/Services/ImageService.cs
using ImageViewer.Models;
using System.IO;

namespace ImageViewer.Services;

public class ImageService
{
    private static readonly string[] SupportedExtensions = 
    {
        ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".tiff", ".tif"
    };

    public List<ImageInfo> LoadImagesFromFolder(string folderPath)
    {
        var imageFiles = new List<ImageInfo>();

        if (!Directory.Exists(folderPath))
            return imageFiles;

        try
        {
            var files = Directory.GetFiles(folderPath)
                .Where(file => SupportedExtensions.Contains(Path.GetExtension(file).ToLowerInvariant()))
                .OrderBy(file => Path.GetFileName(file));

            foreach (var file in files)
            {
                var fileInfo = new FileInfo(file);
                imageFiles.Add(new ImageInfo
                {
                    FilePath = file,
                    FileName = fileInfo.Name,
                    FileSize = FormatFileSize(fileInfo.Length),
                    DateModified = fileInfo.LastWriteTime,
                    Extension = fileInfo.Extension.ToLowerInvariant()
                });
            }
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException($"Error loading images from folder: {ex.Message}", ex);
        }

        return imageFiles;
    }

    private static string FormatFileSize(long bytes)
    {
        string[] sizes = { "B", "KB", "MB", "GB" };
        double len = bytes;
        int order = 0;
        while (len >= 1024 && order < sizes.Length - 1)
        {
            order++;
            len = len / 1024;
        }
        return $"{len:0.##} {sizes[order]}";
    }
}

File: src/ImageViewer-mac/ImageViewer-mac/Services/ImageService.swift
//
//  ImageService.swift
//  ImageViewer-mac
//
//  Created by Michael Hammond on 7/16/25.
//

import Foundation

class ImageService: ObservableObject {
    private static let supportedExtensions = [".png", ".jpg", ".jpeg", ".gif", ".bmp", ".tiff", ".tif", ".mp4", ".heic", ".heif", ".webp"]
    
    func loadImagesFromFolder(_ folderURL: URL) throws -> [ImageInfo] {
        guard folderURL.hasDirectoryPath else {
            throw ImageServiceError.invalidFolder
        }
        
        let fileManager = FileManager.default
        let resourceKeys: [URLResourceKey] = [.isRegularFileKey, .fileSizeKey, .contentModificationDateKey]
        
        guard let enumerator = fileManager.enumerator(at: folderURL.resolvingSymlinksInPath(),
                                                     includingPropertiesForKeys: resourceKeys,
                                                     options: [.skipsHiddenFiles, .skipsSubdirectoryDescendants]) else {
            throw ImageServiceError.cannotEnumerateFiles
        }
        
        var imageInfos: [ImageInfo] = []
        
        for case let fileURL as URL in enumerator {
            let fileExtension = fileURL.pathExtension.lowercased()
            guard Self.supportedExtensions.contains("." + fileExtension) else { continue }
            
            do {
                let resourceValues = try fileURL.resourceValues(forKeys: Set(resourceKeys))
                guard let isRegularFile = resourceValues.isRegularFile, isRegularFile else { continue }
                
                let fileSize = resourceValues.fileSize ?? 0
                let dateModified = resourceValues.contentModificationDate ?? Date()
                
                let imageInfo = ImageInfo(
                    filePath: fileURL.path,
                    fileName: fileURL.lastPathComponent,
                    fileSize: ByteCountFormatter.string(fromByteCount: Int64(fileSize), countStyle: .file),
                    dateModified: dateModified,
                    fileExtension: "." + fileExtension
                )
                
                imageInfos.append(imageInfo)
            } catch {
                print("Error reading file attributes for \(fileURL.path): \(error)")
            }
        }
        
        return imageInfos
    }
}

enum ImageServiceError: LocalizedError {
    case invalidFolder
    case cannotEnumerateFiles
    
    var errorDescription: String? {
        switch self {
        case .invalidFolder:
            return "The selected item is not a valid folder."
        case .cannotEnumerateFiles:
            return "Cannot access files in the selected folder."
        }
    }
}
